### ISSUE TITLE
Experience solid inconsistencies in BCAD

### ISSUE THREAD

**Issue by schoeller** [hakanaktt/acadrust#73]
Dear all,

thanks for this work.

I am trying out writing solids to be openend in ACAD/BCAD. I run

```rust
use acadrust::entities::acis::primitives;
use acadrust::entities::{EntityType, Solid3D};
use acadrust::{CadDocument, DwgWriter};

fn main() -> acadrust::Result<()> {
    let mut doc = CadDocument::new();

    let sat = primitives::build_cylinder([0.0, 0.0, 0.0], 1.0, 2.0);
    let mut solid = Solid3D::from_sat(&sat.to_sat_string());
    solid.point_of_reference = acadrust::types::Vector3::new(0.0, 0.0, 0.0);

    doc.add_entity(EntityType::Solid3D(solid))?;

    DwgWriter::write_to_file("cylinder.dwg", &doc)?;

    let size = std::fs::metadata("cylinder.dwg")
        .map(|m| m.len())
        .unwrap_or(0);
    println!("Wrote cylinder.dwg ({} bytes)", size);

    Ok(())
}
```

when opening the file in BCAD v26 and doing AUDIT inconsistencies are reported. Similar experience I had when writing gen_all_entities_all_versions_dwg.rs

```
Name: AcDbTableStyle(2C)
Value: TextStyle ((0))
```

ACAD did not like both files at all. Hints welcome.

Best

Seb

[cylinder.zip](https://github.com/user-attachments/files/31928111/cylinder.zip)

**Comment by schoeller** [comment]
For testing purposes I did gen_all_entities_all_versions -> dwg. Please find attached source code + sample dwg.

[gen_all_entities_all_versions.zip](https://github.com/user-attachments/files/31965051/gen_all_entities_all_versions.zip)

**Comment by hakanaktt** [comment]
I am heavily working on 3DSolids nowadays for my custom solid modeling engine
Validations are done with BricsCAD and IntelliCAD. I will look into the issue. Maybe include BCAD into my testing environment.

**Comment by schoeller** [comment]
Thank for picking this up.
I just tried with OpenStudioCad compiled at 56e09cd3f580a1b5a9fb4e71935456d23b07e878. A CYLINDER origin 0,0,0 radius 1 height 2 with SAVEAS R2018 fails to open in ACAD 2026/BCAD 26. I have assembled a golden-reference from [LibreDWG](https://github.com/LibreDWG/libredwg) for R2018. It contains dwg, logs, json. Would that be of help for testing? 

[r2018.zip](https://github.com/user-attachments/files/31997528/r2018.zip)

<img width="480" alt="Image" src="https://github.com/user-attachments/assets/b861e7ec-6d33-433c-8296-daa486cb35cf" />

Asking around I have been given these references

https://aps.autodesk.com/developer/overview/objectarx-autocad-sdk
https://images.autodesk.com/adsk/files/autocad_2012_pdf_dxf-reference_enu.pdf
https://www.opendesign.com/files/guestdownloads/OpenDesign_Specification_for_.dwg_files.pdf

**Comment by hakanaktt** [comment]
Versions might differ slightly, OCS' developer is using its own fork and he is sending pull requests to me.
When I directly build the cylinder from acadrust using the code u provided, in current main branch, BCAD 19 and ICAD 15 succesfully opens it w/o problem.
If I understand right, problem might be in OCS' version(cadcodec crate)
Edit: investigating more...might have found something

**Comment by hakanaktt** [comment]
Tablestyle writing might have some issues for 2010+. Will be fixed very soon, if not already.

**Comment by schoeller** [comment]
> Versions might differ slightly, OCS' developer is using its own fork and he is sending pull requests to me. When I directly build the cylinder from acadrust using the code u provided, in current main branch, BCAD 19 and ICAD 15 succesfully opens it w/o problem. If I understand right, problem might be in OCS' version(cadcodec crate) Edit: investigating more...might have found something

Cannot confirm. Pulled your origin. Am on W11 with rustc 1.98.0

```
cargo run --example cylinder_dwg
cargo run --example gen_all_entities_all_versions_dwg
```

**cylinder.dwg:**
_ACAD 2026_
does not like it
<img width="360" alt="Image" src="https://github.com/user-attachments/assets/3a754cbe-5747-434b-bf4e-e6f0b2454da9" />
_BCAD 26_
opens file and shows geometry, then
```
: AUDIT
Fix any errors detected? [Yes/No] <No>: 
Name: AcDbTableStyle(2C)
Value: TextStyle ((0))
Validation: Invalid
Default value: Set to Default

Name: AcDbTableStyle(2C)
Value: TextStyle ((0))
Validation: Invalid
Default value: Set to Default

Name: AcDbTableStyle(2C)
Value: TextStyle ((0))
Validation: Invalid
Default value: Set to Default

91 objects audited
Total errors found during audit 3, fixed 0
```

**gen_all_entities_all_versions.dwg**
_ACAD 2026_
does not like it
<img width="360" alt="Image" src="https://github.com/user-attachments/assets/7a9477ab-256b-4071-9208-71f12476d569" />
_BCAD26_
opens file and shows geometry, then
```
: AUDIT
Fix any errors detected? [Yes/No] <No>: n
Name: AcDbTableStyle(2C)
Value: TextStyle ((0))
Validation: Invalid
Default value: Set to Default

Name: AcDbTableStyle(2C)
Value: TextStyle ((0))
Validation: Invalid
Default value: Set to Default

Name: AcDbTableStyle(2C)
Value: TextStyle ((0))
Validation: Invalid
Default value: Set to Default

Name: AcDbLeader(3F)
Value: Annotation enabled but Id is Null
Validation: Disparity
Default value: Disable annotation

Name: AcDb3dSolid(50)
Value: Modeling operation error:
Data stream is empty
Validation: Invalid
Default value: Removed

Name: AcDbBody(52)
Value: Modeling operation error:
Data stream is empty
Validation: Invalid
Default value: Removed

Name: AcDbRegion(51)
Value: Modeling operation error:
Data stream is empty
Validation: Invalid
Default value: Removed

Name: AcDbShape(42)
Value: Shape Style (font file) is not set
Validation: Invalid
Default value: Removed

Name: AcDbShape(42)
Value: Shape number is not set
Validation: Invalid
Default value: Removed

154 objects audited
Total errors found during audit 9, fixed 0
```

Files attached
[cylinder.zip](https://github.com/user-attachments/files/31999761/cylinder.zip)
[gen_all_entities_all_versions.zip](https://github.com/user-attachments/files/31999762/gen_all_entities_all_versions.zip)

**Comment by hakanaktt** [comment]
1)Fixed the cylinder sample errors regarding TableStyle. pushed it upstream.
2)gen_all_entities_all_versions might be outdated, will run a full audit on those examples now.
3)AutoCAD never likes DWG coming from non Autodesk software. There is a way to make AutoCAD accept these without getting them flagged but afaik it is not legal.

**Comment by schoeller** [comment]
pulled origin and recompiled. test results as follows:

**cylinder.dwg**
_BCAD v18_
OPEN and AUDIT no errors

_BCAD v26_
OPEN and AUDIT no errors

_ACAD 2026_
goes directly into RECOVERY and gives no usable result

**gen_all_entities_all_versions.dwg**
_BCAD v18_
opening message
```
The drawing has proxy objects.

Application name: ACDB_MLEADER_CLASS
Number of proxies: 1, of which 1 entities
```
CMD output
```
General modeling failure
AcDb3dSolid(50)

General modeling failure
AcDbRegion(51)

General modeling failure
AcDbBody(52)
: AUDIT
Fix any errors detected? [Yes/No] <No>:

Name: AcDbLeader(3F)
Value: Annotation enabled but Id is Null
Validation: Disparity
Default value: Disable annotation

Name: AcDbShape(42)
Value: Shape Style (font file) is not set
Validation: Invalid
Default value: Removed

Name: AcDbShape(42)
Value: Shape number is not set
Validation: Invalid
Default value: Removed

170 objects audited
Total errors found during audit 3, fixed 0
```

_BCAD v26_
no opening message, then
```
: AUDIT
Fix any errors detected? [Yes/No] <No>: 
Name: AcDbLeader(3F)
Value: Annotation enabled but Id is Null
Validation: Disparity
Default value: Disable annotation

Name: AcDb3dSolid(50)
Value: Modeling operation error:
Data stream is empty
Validation: Invalid
Default value: Removed

Name: AcDbBody(52)
Value: Modeling operation error:
Data stream is empty
Validation: Invalid
Default value: Removed

Name: AcDbRegion(51)
Value: Modeling operation error:
Data stream is empty
Validation: Invalid
Default value: Removed

Name: AcDbShape(42)
Value: Shape Style (font file) is not set
Validation: Invalid
Default value: Removed

Name: AcDbShape(42)
Value: Shape number is not set
Validation: Invalid
Default value: Removed

154 objects audited
Total errors found during audit 6, fixed 0
```

_ACAD 2026_
goes directly into RECOVERY and gives no usable result

Not being able to use ACAD is quite harsh and should be fixed from user perspective, otherwise one would be required to own a BCAD license in order to be able to open/save through third party.

**Comment by hakanaktt** [comment]
Hopefully soon enough with Open CAD Studio there won't be any need for that.
I am checking why other sample gives that many errors, will be fixed today dutifully.

**Comment by hakanaktt** [comment]
Probably ACAD has stricter checks and there are some false flags. I will work on preventing entering RECOVERY mode while using AutoCAD at least. That's at least I can do, some formats are giving me this error so far(1021,1027,1032) on ACAD 2027. By version 0.6 I believe acadrust will be fully capable of writing ACAD valid DWG's.

**Comment by schoeller** [comment]
> 1021,1027,1032

Not knowing what I do. Searching for numbers in LibreDWG repo gives

dynapi.c
{ "AECC_DISP_REP_SECTION_CORRIDOR_POINT_LABEL_GROUP", DWG_TYPE_AECC_DISP_REP_SECTION_CORRIDOR_POINT_LABEL_GROUP /*(1021)*/, NULL, 0 },	/* 300 */
{ "AECC_DISP_REP_SECTION_PROJECTION", DWG_TYPE_AECC_DISP_REP_SECTION_PROJECTION /*(1027)*/, NULL, 0 },	/* 306 */
{ "AECC_DISP_REP_SECTION_VIEW_QUANTITY_TAKEOFF_TABLE", DWG_TYPE_AECC_DISP_REP_SECTION_VIEW_QUANTITY_TAKEOFF_TABLE /*(1032)*/, NULL, 0 },	/* 311 */

alldxf_0.inc
{ "TABLESTYLE", "test/test-data/2010/ES2_S.dxf", 0xDC, /* 1021 */
{ "TABLESTYLE", "test/test-data/example_2018.dxf", 0x29B, /* 1027 */
{ "TABLESTYLE", "test/test-data/2004/fr05_f101_ref.dxf", 0x483, /* 1032 */

Suppose that you are receiving this as an error. Will have to review how error numbering is handled there. Don't count on me.

**Comment by hakanaktt** [comment]
I am still working on it. Validating every entity structure on every possible version directly on ACAD 2027 and BCAD20. Preparing a giant atlas drawing that showcases all entities available. It includes DXF/DWG and DXB targets.

**Comment by schoeller** [comment]
> I am still working on it. Validating every entity structure on every possible version directly on ACAD 2027 and BCAD20. Preparing a giant atlas drawing that showcases all entities available.

Wow, is that what libredwg people have been trying for years? Hope you'll find the end of the thread...

**Comment by hakanaktt** [comment]
Exactly, especially DWG is where hardness arise. That's also the reason you wont find AC1021(2007 DWG) writing support anywhere on open source libraries. It is all about reverse engineering. Thanks we have AI on our side this time and it can analyze code better and understand differences better than any human being. There still will be some edge cases but hopefully acadrust will be a much more reliable library at 0.6.0.

**Comment by schoeller** [comment]
Just in case this provides useful to you: I forward the [golden reference R2018](https://github.com/user-attachments/files/31997528/r2018.zip) from their unit-tests in a post above. It contains DWG, JSON, log per entity. It is quite explicit. Maybe AI is able to end-to-end test automagically and repair iteratively differences, or magically extract a ruleset. Dunno how that would work in detail, but an idea. Example2018.dwg contains all sorts of object including a solid. It all loads smoothly into ACAD.
It seems possible to generate these for 2004, 2007, 2010, 2013, 2018.

JSON
<img width="300" alt="Image" src="https://github.com/user-attachments/assets/2b7454fe-1ade-4bf0-87ba-017e53a7f540" />

log
<img width="300" alt="Image" src="https://github.com/user-attachments/assets/32fafbe5-b7d9-43d6-b1e7-362438f501db" />

example2018.dwg
<img width="600" alt="Image" src="https://github.com/user-attachments/assets/78f01ba1-49a5-44c9-918d-2a8fc29ff790" />

**Comment by hakanaktt** [comment]
0.5.5 should fix most of these issues
It is out

**Comment by hakanaktt** [comment]
Check https://github.com/hakanaktt/acadrust/blob/main/src%2Fdocs%2Fentity_status_matrix.md

**Comment by schoeller** [comment]
@hakanaktt thanks for the extra shifts. I was trying from the other side. Dunno whether its confusing or misleading experiment.

1) I asked AI to scan libredwg golden files R2018 and to develop a unique machine readable write-logic atlas for R2018. It came up with 10.9MB JSON file and confirmed it would be unique (r2018_write_logic_atlas.json).
2) I then asked it to take the write-logic atlas R2018 and scan entire acadrust codebase for an audit comparing the write-logic sequence. It came up with a report (AUDIT_R2018.zip). I asked to verify the report to be unique, which it it stated, but I do not trust the machine. So whether mentions marked as CRITICAL are truly critical I cannot tell.

In case this provides useful to you I might try generating a 
3) unique machine readable read-logic atlas R2018
4) audit for read-logic R2018
5) etc.

[r2018_write_logic_atlas.zip](https://github.com/user-attachments/files/32056603/r2018_write_logic_atlas.zip)
[AUDIT_R2018.zip](https://github.com/user-attachments/files/32056621/AUDIT_R2018.zip)

<img width="600" alt="Image" src="https://github.com/user-attachments/assets/e734306b-8275-4bb2-aab7-b69e8f2fc2ab" />

**Comment by hakanaktt** [comment]
@schoeller Thanks for going at it from this angle, and for the offer to keep going. I went through `AUDIT_R2018.md` in detail. Short version: the idea is sound, but this particular report is dominated by tool bugs rather than acadrust bugs, and I'd rather tell you that now than have you spend the effort on a read-logic atlas that would hit the same wall.

Your instinct not to trust the machine was right. Here are the four systematic problems, each checked against the source.

**1. The entity preamble is never inlined.** This is the big one. The report says `write_arc` emits 13 ops — that is exactly the body of `write_arc` in `src/io/dwg/dwg_stream_writers/object_writer/entities.rs`, with nothing pulled in from `entity_preamble` → `write_common_entity_data` (`object_writer/common.rs`), even though the stated method claims helpers are recursively inlined. So `entmode`, `num_reactors`, `is_xdic_missing`, `has_ds_data`, `ltype_flags`, `plotstyle_flags`, `material_flags`, the three visual-style bits, `color.name`, `linewt` are all reported CRITICAL-missing while being written a few call frames up. That single failure accounts for the bulk of the 3889 CRITICAL findings.

You can see the alignment collapse in the same table: `center.x`, `center.y`, `radius`, `start_angle`, `end_angle` are listed as missing, and then the *same* values show up as LOW "extra in acadrust" a few rows below. Nothing is missing there, the two sequences just lost sync once the preamble ops vanished from one side.

**2. Version filtering isn't applied.** `flag_r11`, `elevation_r11` and `thickness_r11` are pre-R13 fields, flagged CRITICAL for AC1032. Likewise every table control demands both `bit_write_RS num_entries` (pre-R2000) *and* `bit_write_BS num_entries` (R2000+) simultaneously — mutually exclusive branches counted as two missing fields. Step 8 of the method says the golden ops get re-filtered for R2018; that clearly isn't happening.

**3. Shared dispatchers are compared against their entire match-arm set.** Every `ACSH_*` type reports 1236 acadrust ops, every `ASSOC*` 849, every `write_class_object` type 518, every context-data object 356. Those are the totals for the whole `match` in `write_dynamic_block` / `write_associative_object` / etc., counted once per type instead of resolving to the arm that type actually takes. That is where the 170331 LOW findings come from — it's the same dispatcher body re-counted a few hundred times.

**4. The tool flags its own scaffolding.** `bit_write_MS len`, `bit_write_UMC handle_bits` and `bit_write_CRC` appear as "extra in acadrust" in all 304 types, but step 9 of the method says the tool prepends and appends those itself. Similarly `bit_copy_chain`, `obj_flush_hdlstream` and `bit_write_RS _data_size` are LibreDWG's stream plumbing, not per-entity fields — acadrust does the equivalent in the merged writer and in `register_object`, outside the per-type function the tool is reading.

The INFO tier is the same drift as (1): `num_reactors` paired against `style_type` only because the preamble went missing on the acadrust side first.

**The one item I do want to look at:** VISUALSTYLE. `write_visual_style` gates the legacy property block behind `!r2010_plus()` and takes a different property-array path for R2010+. The audit's "missing" list for it is drawn from the pre-R2010 golden path, so the report doesn't establish anything, but I haven't hand-verified that the R2010+ branch matches AC1032 field for field. That one's on my list regardless.

**What would actually work.** The problem is that this compares source code to source code. To get a meaningful diff that way you have to resolve version gating, helper inlining and enum dispatch correctly, and here none of the three worked — so the output can't distinguish a real ordering bug from an unexpanded function call.

A byte-level differential avoids the whole class of error: take one entity, write it with acadrust and with LibreDWG at the same version, decode both bitstreams, and diff the decoded field sequence by position. Any disagreement is real by construction — no inlining to get right, no conditions to re-evaluate, no false positives. That's the direction I'd put effort into, and it's roughly what I'm doing manually for the entity/version matrix right now.

So: please don't build the read-logic atlas on this method. If you're up for it, a per-field bitstream dump for a handful of entity types out of the R2018 golden files would be genuinely useful to me — that's the reference side of exactly the differential above.

Either way, thanks. The AUDIT sent me back through the common-entity preamble with fresh eyes, which was worth the read on its own.

**Comment by schoeller** [comment]
Example bitdump with patched libredwg for test-files Arc, Constraints, Donut in R2018. Is this of any use? If no, please indicate exact required format, if yes I may generated the lot upon your word.

[r2018_Arc.zip](https://github.com/user-attachments/files/32158161/r2018_Arc.zip)
[r2018_Constraints.zip](https://github.com/user-attachments/files/32158189/r2018_Constraints.zip)
[r2018_Donut.zip](https://github.com/user-attachments/files/32158196/r2018_Donut.zip)

**Comment by hakanaktt** [comment]
These bitdumps are of no use, as I told you before. You can compare same DWG exported from acadrust and libreDWG or you can roundtrip a LibreDWG file(acadrust reads the DWG file and writes back in same format) then compare. This would be actually helpful.

**Comment by hakanaktt** [comment]
If you find an issue with a file that you got out of acadrust i, feel free to share here in a new issue. I would be glad to check out.

**Comment by schoeller** [comment]
@hakanaktt : FYI have developed a [gold (LibreDWG) vs silver (Acadrust) test](https://github.com/schoeller/cadcodec/blob/0f2b6c1b1931a38b3e7bb5120ab4d1edef23fcfb/tests/gold_harness/README.md). Dunno how far I will get, but still struggling with OPEN and AUDIT in ACAD/BCAD, also in latest OCS version. Will let you know when ready for pull.

**Comment by hakanaktt** [comment]
I wouldn't call libreDWG a golden corpus for 3D entities but for 2D ones it might be helpful. Plus we can decode any file from ACAD(DWG/DXF whatsoever) so instead of libredwg corpus, I would advise a ACAD/BCAD one that would be more helpful.

**Comment by schoeller** [comment]
> I wouldn't call libreDWG a golden corpus for 3D entities but for 2D ones it might be helpful. Plus we can decode any file from ACAD(DWG/DXF whatsoever) so instead of libredwg corpus, I would advise a ACAD/BCAD one that would be more helpful.

Would you know of a golden fileset ACAD/BCAD made?

**Comment by schoeller** [comment]
> [@hakanaktt](https://github.com/hakanaktt) : FYI have developed a [gold (LibreDWG) vs silver (Acadrust) test](https://github.com/schoeller/cadcodec/blob/0f2b6c1b1931a38b3e7bb5120ab4d1edef23fcfb/tests/gold_harness/README.md). Dunno how far I will get, but still struggling with OPEN and AUDIT in ACAD/BCAD, also in latest OCS version. Will let you know when ready for pull.

Finished in [gold-vs-silver](https://github.com/schoeller/cadcodec/tree/gold-vs-silver) branch with `cargo test` passing. Main part is in [tests/gold_harness](https://github.com/schoeller/cadcodec/tree/gold-vs-silver/tests/gold_harness). It is heavy, but (with the help of many tokens) has managed to zero in gold-vs-silver (libredwg vs. acadrust). If there was a golden ref from ACAD itself I would have done differently, but I have no dataset. [gen_all_entities_all_versions_dwg.rs](https://github.com/schoeller/cadcodec/blob/gold-vs-silver/examples/gen_all_entities_all_versions_dwg.rs) as well as [cylinder](https://github.com/hakanaktt/acadrust/issues/73#issue-5378968493) example opens flawlessly in BCAD.
When time permits please kindly review. PR may be generated upon your call.

**Comment by schoeller** [comment]
> > [@hakanaktt](https://github.com/hakanaktt) : FYI have developed a [gold (LibreDWG) vs silver (Acadrust) test](https://github.com/schoeller/cadcodec/blob/0f2b6c1b1931a38b3e7bb5120ab4d1edef23fcfb/tests/gold_harness/README.md). Dunno how far I will get, but still struggling with OPEN and AUDIT in ACAD/BCAD, also in latest OCS version. Will let you know when ready for pull.
> 
> Finished in [gold-vs-silver](https://github.com/schoeller/cadcodec/tree/gold-vs-silver) branch with `cargo test` passing. Main part is in [tests/gold_harness](https://github.com/schoeller/cadcodec/tree/gold-vs-silver/tests/gold_harness). It is heavy, but (with the help of many tokens) has managed to zero in gold-vs-silver (libredwg vs. acadrust). If there was a golden ref from ACAD itself I would have done differently, but I have no dataset. [gen_all_entities_all_versions_dwg.rs](https://github.com/schoeller/cadcodec/blob/gold-vs-silver/examples/gen_all_entities_all_versions_dwg.rs) as well as [cylinder](https://github.com/hakanaktt/acadrust/issues/73#issue-5378968493) example opens flawlessly in BCAD. When time permits please kindly review. PR may be generated upon your call.

MLEADER constructed in OCS fails to load in BCAD. Started investigating and fixing in [gold-vs-silver](https://github.com/schoeller/cadcodec/tree/gold-vs-silver/tests/gold_harness) as permitted. Starting out with 94533a344f90c72e55f55f8668026d669118fa07 for MLEADER. Will continue with other entities generated by [gen_all_entities_all_versions_dwg.rs](https://github.com/schoeller/cadcodec/blob/gold-vs-silver/examples/gen_all_entities_all_versions_dwg.rs) confirmed to load in BCAD.

**Comment by hakanaktt** [comment]
@schoeller please open a new issue related to the file. And please share the problematic file too. Will be glad to check.

**Comment by schoeller** [comment]
> [@schoeller](https://github.com/schoeller) please open a new issue related to the file. And please share the problematic file too. Will be glad to check.

Cannot reproduce with OCS 2026.38. Sorry for the fuzz. Grpf. Will go through the list from [gen_all_entities_all_versions_dwg.rs](https://github.com/schoeller/cadcodec/blob/gold-vs-silver/examples/gen_all_entities_all_versions_dwg.rs) and share as issues in case of trouble accordingly.

**Comment by schoeller** [comment]
@hakanaktt 
have expanded gold-harness test to custom entities in gold-vs-silver branch. am now with SOLIDHISTORY in solids and generated custom gold files (chamfer, fillet, pyramid, box, cylinder etc.). Will try it with OCS.
